Improve html docs (#100)
yewstack/docs: src/concepts/html/README.md

description: The procedural macro for generating HTML and SVG
---

> # Using `html!`

Reviewer: Yes! I think we should adopt this as a convention accross all our documentation (using links to the relevant items in the API docs).

The `html!` macro allows you to write HTML and SVG code declaratively. It is similar to JSX \(a Javascript extension which allows you to write HTML code inside of Javascript\).

**Important notes**

1. The `html!` macro only accepts one root html node \(you can counteract this by [using fragments or iterators](lists.md)\)
2. An empty `html! {}` invocation is valid and will not render anything

> 3. Literals in element content must always be quoted and wrapped in braces

Reviewer: Not sure if "in element content" helps here?

Author: Any better suggestions? I thought a bit but that didn't help ;-)

Reviewer: Could we just say?
```suggestion
3. Literals must always be quoted and wrapped in braces
```

Author: That was what confused me because literal values passed to attributes do not need braces.

Reviewer: Oh right. That makes sense. I guess 
```suggestion
3. Literals inside of tags must always be quoted and wrapped in braces (e.g. `<p>{"hello world"}</p>`)
```

  * `html! { "Hello, World" }`
  * `html! { <div>{ "Hell, World" }</div> }`
  * `html! { <div>{ String::from("foo") + "bar" }</div>`

> 4. Quoted attribute values are taken literally

Reviewer: ```suggestion
4. Quoted attribute values are used as-is (i.e. their values are set at compile-time and do not change at run-time)
```

  * `html! { <div> id="bar"</div> }`

> 5. Unquoted attribute values are interpreted as expressions

Reviewer: ```suggestion
5. Unquoted attribute values are interpreted as expressions (which must be valid Rust expressions)
```

Author: What about just 
```
Unquoted attribute values are interpreted as Rust expressions
```
I'd say that they need to be valid does not need to be mentioned.

Reviewer: I think it clears any confusion as to whether expressions feature any syntax extensions upon those of regular Rust.

description: The procedural macro for generating HTML and SVG
---

# Using [`html!`](https://docs.rs/yew/0.16.2/yew/macro.html.html)

The [`html!`](https://docs.rs/yew/0.16.2/yew/macro.html.html) macro allows you to write HTML and SVG code declaratively. It is similar to JSX \(a Javascript extension which allows you to write HTML code inside of Javascript\).

**Important notes**

> - The [`html!`](https://docs.rs/yew/0.16.2/yew/macro.html.html) macro only accepts one root html node (you can counteract this by [using fragments or iterators](lists.md))

Reviewer: ```suggestion
- The [`html!`](https://docs.rs/yew/0.16.2/yew/macro.html.html) macro only accepts one root HTML node (you can overcome this by [using fragments or iterators](lists.md))
```

- An empty `html! {}` invocation is valid and will not render anything.

> - Literals in element content must always be quoted and wrapped in braces (in contrary to attribute values - see below)

Reviewer: ```suggestion
- Literals in element content must always be quoted and wrapped with braces (this is different to attribute values - see below)
```

  * `html! { "Hello, World" }`
  * `html! { <div>{ "Hell, World" }</div> }`
  * `html! { <div>{ String::from("foo") + "bar" }</div>`

> - Quoted attribute values are taken literally. The value is set at compile-time and does not change at runtime.

Reviewer: I think I made this change (so my fault here – just want to be consistent).
```suggestion
- Quoted attribute values are taken literally. The value is set at compile-time and does not change at run-time.
```

  * `html! { <div> id="bar"</div> }`
- Unquoted attribute values are interpreted as expressions and therefore have to be valid Rust expressions.
  * `let foo = "bar"; html! { <div id=foo></div> }`
  * `html! { <div id=String::from("foo") + "bar"></div> }`

{% hint style="info" %}

> The [`html!`](https://docs.rs/yew/0.16.2/yew/macro.html.html) macro can reach easily the default recursion limit of the compiler. It is advised to bump its value if you encouter compilation errors. Use an attribute like `#![recursion_limit="1024"]` to bypass the problem. See the [official documentation](https://doc.rust-lang.org/reference/attributes/limits.html#the-recursion_limit-attribute) and [this Stack Overflow question](https://stackoverflow.com/questions/27454761/what-is-a-crate-attribute-and-where-do-i-add-it) for details.

Reviewer: ```suggestion
The [`html!`](https://docs.rs/yew/0.16.2/yew/macro.html.html) macro can easily exceed the default recursion limit of the compiler. It is advised to bump its value if you encouter compilation errors. Use an attribute like `#![recursion_limit="1024"]` to bypass the problem. See the [official documentation](https://doc.rust-lang.org/reference/attributes/limits.html#the-recursion_limit-attribute) and [this Stack Overflow question](https://stackoverflow.com/questions/27454761/what-is-a-crate-attribute-and-where-do-i-add-it) for details.
```

Author: I can't really see what exactly the change was here. 

What do you think about introducing a fix line length limit?
It's much easier to compare lines that are not broken over several lines.

I don't know what IDE you're using but in Emacs a keyboard shortcut realigns text to fit the defined width.

I just saw that `latest` works in links instead of a concrete version number. 
I think that's better than to update all links with each version increment.

I added a link in another file. I intended to create a new PR for that but missed that I had a type when I tried to switch the branch so it landed in this branch as well.

Reviewer: "can reach easily" => "can easily exceed"

Reviewer: I'll create an issue (and PR) about line lengths.

description: The procedural macro for generating HTML and SVG
---

# Using [`html!`][yew_macro_html]

The [`html!`][1] macro allows you to write HTML and SVG code declaratively. It is similar to JSX \(a Javascript extension which allows you to write HTML code inside of Javascript\).

**Important notes**

- The [`html!`][yew_macro_html] macro only accepts one root HTML node (you can overcome this by [using fragments or iterators](lists.md))
- An empty `html! {}` invocation is valid and will not render anything.
- Literals inside of tags must always be quoted and wrapped with braces (this is different to attribute values - see below)
  * `html! { "Hello, World" }`
  * `html! { <div>{ "Hell, World" }</div> }`
  * `html! { <div>{ String::from("foo") + "bar" }</div>`
- Quoted attribute values are taken literally. The value is set at compile-time and does not change at run-time.
  * `html! { <div> id="bar"</div> }`
- Unquoted attribute values are interpreted as expressions and therefore have to be valid Rust expressions.
  * `let foo = "bar"; html! { <div id=foo></div> }`

>   * `html! { <div id=String::from("foo") + "bar"></div> }`

Reviewer 2: This is a bit confusing because literals are also expressions. I'd prefer if we left these notes out

> - HTML tags names need to start with a lowercase letter. Besides that every valid HTML tag name is allowed. If the tag name starts with an uppercase letter it is interpreted as component name and attributes are passed as component properties.

Reviewer 2: ```suggestion
- Tag names that begin with a lowercase letter will be treated as the name of a DOM elements.
- Tag names that begin with an uppercase letter will be treated as a type implementing the `Component` trait and the attributes will be passed as component properties.
```
